File: src/types/certificate.ts
import { z } from 'zod'

export const qualityCompositionSchema = z.object({
  totalOnions: z.number().nullable().optional(),
  gradeAPercent: z.number().nullable().optional(),
  ursPercent: z.number().nullable().optional(),
  undersizedCount: z.number().nullable().optional(),
  rottenDamagedCount: z.number().nullable().optional(),
  sproutedCount: z.number().nullable().optional(),
})

export type QualityComposition = z.infer<typeof qualityCompositionSchema>

export const certificateSchema = z.object({
  id: z.string(),
  inspectionId: z.string(),
  grade: z.string(),
  issuedAt: z.string(),
  batchLabel: z.string(),
  qrToken: z.string(),
  inspectorName: z.string().nullable().optional(),
  batchId: z.string().nullable().optional(),
  procurementCentre: z.string().nullable().optional(),
  specification: z.string().nullable().optional(),
  sampleSize: z.number().nullable().optional(),
  confidence: z.number().nullable().optional(),
  defectSummary: z.string().nullable().optional(),
  aiGrade: z.string().nullable().optional(),
  officerGrade: z.string().nullable().optional(),
  overrideCount: z.number().optional(),
  dualAssessment: z.record(z.string(), z.any()).nullable().optional(),
  whyThisGrade: z.record(z.string(), z.any()).nullable().optional(),
  standardsMatrix: z.array(z.record(z.string(), z.any())).nullable().optional(),
  qualityComposition: qualityCompositionSchema.nullable().optional(),
  auditTimeline: z
    .array(
      z.object({
        event: z.string(),
        time: z.string(),
      }),
    )
    .nullable()
    .optional(),
})

export const auditEventSchema = z.object({
  event: z.string(),
  time: z.string(),
})

export type AuditEvent = z.infer<typeof auditEventSchema>

export type Certificate = z.infer<typeof certificateSchema>

export const verificationResultSchema = z.object({
  valid: z.boolean(),
  certificate: certificateSchema.nullable().optional(),
  message: z.string(),
  auditTimeline: z.array(auditEventSchema).nullable().optional(),
  whyThisGrade: z.record(z.string(), z.any()).nullable().optional(),
  standardsMatrix: z.array(z.record(z.string(), z.any())).nullable().optional(),
  farmerTransparency: z.record(z.string(), z.any()).nullable().optional(),
  qualityComposition: qualityCompositionSchema.nullable().optional(),
})

export type VerificationResult = z.infer<typeof verificationResultSchema>

File: backend/app/schemas/certificate.py
from pydantic import BaseModel, ConfigDict


class AuditEvent(BaseModel):
    event: str
    time: str


class QualityComposition(BaseModel):
    model_config = ConfigDict(exclude_none=True)

    totalOnions: int | None = None
    gradeAPercent: float | None = None
    ursPercent: float | None = None
    undersizedCount: int | None = None
    rottenDamagedCount: int | None = None
    sproutedCount: int | None = None


class Certificate(BaseModel):
    model_config = ConfigDict(exclude_none=True)

    id: str
    inspectionId: str
    grade: str
    issuedAt: str
    batchLabel: str
    qrToken: str
    inspectorName: str | None = None
    batchId: str | None = None
    procurementCentre: str | None = None
    specification: str | None = None
    sampleSize: int | None = None
    confidence: float | None = None
    defectSummary: str | None = None
    auditTimeline: list[AuditEvent] | None = None
    aiGrade: str | None = None
    officerGrade: str | None = None
    overrideCount: int = 0
    dualAssessment: dict[str, object] | None = None
    whyThisGrade: dict[str, object] | None = None
    standardsMatrix: list[dict[str, object]] | None = None
    qualityComposition: QualityComposition | None = None


class VerificationResult(BaseModel):
    model_config = ConfigDict(exclude_none=True)

    valid: bool
    message: str
    certificate: Certificate | None = None
    auditTimeline: list[AuditEvent] | None = None
    whyThisGrade: dict[str, object] | None = None
    standardsMatrix: list[dict[str, object]] | None = None
    qualityComposition: QualityComposition | None = None
    farmerTransparency: dict[str, object] | None = None
